sky/spot/recovery_strategy.py

```diff
@@ -142,39 +156,66 @@ def _launch(self, max_retry=3, raise_on_failure=True) -> Optional[float]:
                 # code.
                 logger.info(
                     f'Failed to launch the spot cluster with error: {e}')
-
-            cluster_status, _ = backend_utils.refresh_cluster_status_handle(
-                self.cluster_name, force_refresh=True)
-            if cluster_status == global_user_state.ClusterStatus.UP:
-                # Wait the job to be started
+                retry_launch = True
+                exception = e
+
+            status = None
+            job_checking_retry_cnt = 0
+            # Check the job status until it is not in initialized status.
+            while True:
+                job_checking_retry_cnt += 1
+                if job_checking_retry_cnt >= spot_utils.MAX_JOB_CHECKING_RETRY:
+                    # Avoid the infinite loop, if any bug happens.
+                    retry_launch = True
+                    logger.info(
+                        'Failed to get the job status, due to unexpected '
+                        'job submission error.')
+                    break
+                cluster_status, _ = backend_utils.refresh_cluster_status_handle(
+                    self.cluster_name, force_refresh=True)
+                if cluster_status != global_user_state.ClusterStatus.UP:
+                    # The cluster can be preempted before the job is launched.
+                    # In this case, we will terminate the cluster and retry
+                    # the launch.
+                    retry_launch = True
+                    logger.info(
+                        'The cluster is preempted before the job starts.')
+                    break
+                # TODO(zhwu): The preemption can still happen right before this
+                # line, but it is very unlikely.
                 status = spot_utils.get_job_status(self.backend,
```

Comment:
Ditto - Can this function raise an uncaught exception if the cluster is preempted right before or during this call? My local branch added a try-catch around this. But I lost the logs to verify whether the except block would be triggered.  But an overnight run doesn't return any FAILED_CONTROLLER anymore. Seems safer to add?

Reply:
Good point! Added. Thanks!